### ISSUE TITLE
Support for Array of Struct

### ISSUE THREAD

**Issue by heyvivek31** [confluentinc/ksql#3135]
Hi Team,
We are working on kSQL streams and here we need to filter array of struct fields but there is no such kSQL function to help here. 

> ksql> describe PRODUCT_STREAM;
> 
> Name                 : PRODUCT_STREAM
>  Field        | Type                                                                                                                                                                             
>  ROWTIME      | BIGINT           (system)                                                                                                                                                        
>  ROWKEY       | VARCHAR(STRING)  (system)                                                                                                                                                        
>  `PRODUCTPARAMS | STRUCT<PRODUCTLIST ARRAY<STRUCT<PRODUCTNAME VARCHAR(STRING), PRODUCTCODE VARCHAR(STRING), PRODUCTDESC VARCHAR(STRING)>>> `
>  PRODUCTID      | VARCHAR(STRING) 

I wanted to do something like 

> SELECT PRODUCTID, PRODUCTPARAMS FROM PRODUCT_STREAM WHERE ARRAYCONTAINS(PRODUCTPARAMS->PRODUCTNAME, 'Laptop');

We are getting below error and Looks like ARRAYCONTAINS doesn't accept ARRAY of STRUCT data type .

_Statement: SELECT PRODUCTPARAMS->PRODUCTLIST FROM PRODUCT_STREAM WHERE ARRAYCONTAINS(PRODUCTPARAMS->PRODUCTLIST, 'Laptop');
Caused by: Function 'ARRAYCONTAINS' does not accept parameters of types:[ARRAY,
	VARCHAR(STRING)]_

We then decided to write our own UDF function but still no luck.
```
@UdfDescription(name = "EXTRACTLIST", description = "Loop through Array and filter the data")
public class DataFilter {

@Udf(description = "Loop through Array and filter the data")
public List<Struct> productDataUdf(@UdfParameter(schema = "STRUCT<PRODUCTLIST ARRAY<STRUCT<PRODUCTNAME VARCHAR(STRING), PRODUCTCODE VARCHAR(STRING), PRODUCTDESC VARCHAR(STRING)>>>") final Struct installedAppsStruct){
	//Looping through each item and filter out some items
}
}
```

```
[2019-07-25 13:40:27,285] INFO Adding UDF name='EXTRACTLIST' from path=/opt/confluent-5.3.0/etc/ksql/ext/ProductData-0.0.1-SNAPSHOT-jar-with-dependencies.jar (io.confluent.ksql.function.UdfLoader:213)
[2019-07-25 13:40:27,298] INFO Adding function EXTRACTLIST for method public java.util.List dataFilter.DataFilter.productDataUdf(org.apache.kafka.connect.data.Struct) (io.confluent.ksql.function.UdfLoader:247)
[2019-07-25 13:40:27,383] WARN Failed to add UDF to the MetaStore. name=EXTRACTLIST method=public java.util.List dataFilter.DataFilter.productDataUdf(org.apache.kafka.connect.data.Struct) (io.confluent.ksql.function.UdfLoader:222)
io.confluent.ksql.util.KsqlException: Could not load UDF method with signature: public java.util.List dataFilter.DataFilter.productDataUdf(org.apache.kafka.connect.data.Struct)
	at io.confluent.ksql.function.UdfLoader.getReturnType(UdfLoader.java:373)
	at io.confluent.ksql.function.UdfLoader.addFunction(UdfLoader.java:280)
	at io.confluent.ksql.function.UdfLoader.lambda$handleUdfAnnotation$8(UdfLoader.java:217)
	at io.github.lukehutch.fastclasspathscanner.scanner.ScanSpec$9.lookForMatches(ScanSpec.java:1390)
	at io.github.lukehutch.fastclasspathscanner.scanner.ScanSpec.callMatchProcessors(ScanSpec.java:696)
	at io.github.lukehutch.fastclasspathscanner.FastClasspathScanner.scan(FastClasspathScanner.java:1606)
	at io.github.lukehutch.fastclasspathscanner.FastClasspathScanner.scan(FastClasspathScanner.java:1678)
	at io.github.lukehutch.fastclasspathscanner.FastClasspathScanner.scan(FastClasspathScanner.java:1704)
	at io.confluent.ksql.function.UdfLoader.loadUdfs(UdfLoader.java:145)
	at io.confluent.ksql.function.UdfLoader.lambda$load$2(UdfLoader.java:115)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at io.confluent.ksql.function.UdfLoader.load(UdfLoader.java:115)
	at io.confluent.ksql.rest.server.KsqlRestApplication.buildApplication(KsqlRestApplication.java:473)
	at io.confluent.ksql.rest.server.KsqlRestApplication.buildApplication(KsqlRestApplication.java:441)
	at io.confluent.ksql.rest.server.KsqlServerMain.createExecutable(KsqlServerMain.java:94)
	at io.confluent.ksql.rest.server.KsqlServerMain.main(KsqlServerMain.java:59)
Caused by: io.confluent.ksql.util.KsqlException: Type inference is not supported for: class org.apache.kafka.connect.data.Struct
	at io.confluent.ksql.util.SchemaUtil.handleParametrizedType(SchemaUtil.java:378)
	at io.confluent.ksql.util.SchemaUtil.lambda$getSchemaFromType$5(SchemaUtil.java:158)
	at io.confluent.ksql.util.SchemaUtil.getSchemaFromType(SchemaUtil.java:158)
	at io.confluent.ksql.util.SchemaUtil.getSchemaFromType(SchemaUtil.java:153)
	at io.confluent.ksql.util.SchemaUtil.handleParametrizedType(SchemaUtil.java:371)
	at io.confluent.ksql.util.SchemaUtil.lambda$getSchemaFromType$5(SchemaUtil.java:158)
	at io.confluent.ksql.util.SchemaUtil.getSchemaFromType(SchemaUtil.java:158)
	at io.confluent.ksql.util.SchemaUtil.getSchemaFromType(SchemaUtil.java:153)
	at io.confluent.ksql.function.UdfLoader.getReturnType(UdfLoader.java:366)
	... 26 more
```

Could you please help here ?

**Comment by big-andy-coates** [comment]
Hi @heyvivek31 

Yes, `ARRAYCONTAINS` is a bit of a weird inbuilt function. We'll need to address that at some point. Until then your approach of writing your own UDF is the way to go.

To do what you want you'll need either the new `@Udf(schema='ARRAY<STRUCT<...>>')` or maybe the new `@UdfSchemaProvider` functionality. This is available in master branch and will be released in the upcoming 5.4.

In the interest of keeping our issue backlog tidy I'm going to close this issue. Please feel free to reopen with more details if the answer does not fully answer your question of if you have a follow up question. Thanks for using KSQL!

**Comment by simoncas** [comment]
Hi @big-andy-coates 

I need to write an UDAF to aggregate some Structs S into an array of Struct S.
But if I understand well your previous reply, I need to wait on the upcoming KSQL 5.4 release to be able to use the new features you mention ? I'm using KSQLDB. 

Thanks in advance.

**Comment by rbk1993** [comment]
> We are working on kSQL streams and here we need to filter array of struct fields but there is no such kSQL function to help here.

Hi,

I tried this workaround with KSQL 5.4 : 
I used the [EXPLODE ](https://docs.confluent.io/current/ksql/docs/developer-guide/syntax-reference.html#table-functions) table function, to create one separate record for each entry of the array of structs. Then, I created a STREAM that filters on the struct internal values. (e.g. myStruct->value1 = 'hello').
The problem is that if my array contains two entries that respect the filter (myStruct -> value1 = 'hello'), I end with the two exact same rows in the STREAM (a duplicate). I'm stuck to eliminate one of these rows.
I'm not sure the EXPLODE function can help me further there. And ARRAYCONTAINS doesn't support structs apparently. So, we're obliged to write our own UDF for now ?

Thank you,

**Comment by rbk1993** [comment]
> Hi @heyvivek31
> 
> Yes, `ARRAYCONTAINS` is a bit of a weird inbuilt function. We'll need to address that at some point. Until then your approach of writing your own UDF is the way to go.
> 
> To do what you want you'll need either the new `@Udf(schema='ARRAY<STRUCT<...>>')` or maybe the new `@UdfSchemaProvider` functionality. This is available in master branch and will be released in the upcoming 5.4.
> 
> In the interest of keeping our issue backlog tidy I'm going to close this issue. Please feel free to reopen with more details if the answer does not fully answer your question of if you have a follow up question. Thanks for using KSQL!

Hi @big-andy-coates 

What are the Java packages to import for a @UdfParameter that is ARRAY<STRUCT<...>> ?
I attempted to user Java.util.List<Struct> (with Struct imported from org.apache.kafka.common.protocol.types.Struct) but this doesn't seem to fit.
